aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -2,331 +2,388 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::{
-    bounded_math::{code_invariant_error, expect_ok, BoundedMath, SignedU128},
-    delta_math::{merge_data_and_delta, merge_two_deltas, DeltaHistory},
-    types::{AggregatorID, SnapshotToStringFormula},
+    bounded_math::code_invariant_error,
+    delta_change_set::DeltaOp,
+    types::{AggregatorID, AggregatorValue, SnapshotToStringFormula},
 };
 use move_binary_format::errors::PartialVMResult;
 
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub enum AggregatorChange {
-    Data {
-        value: u128,
-    },
-    StringData {
-        value: Vec<u8>,
-    },
+pub enum AggregatorApplyChange {
     AggregatorDelta {
-        delta: SignedU128,
-        max_value: u128,
-        history: DeltaHistory,
+        delta: DeltaOp,
     },
-    /// Value of delta is:
-    /// formula(value of base_aggregator at the beginning of the transaction + delta)
+    /// Value is:
+    /// (value of base_aggregator at the beginning of the transaction + delta)
     SnapshotDelta {
-        delta: SignedU128,
         base_aggregator: AggregatorID,
+        delta: DeltaOp,
+    },
+    /// Value is:
+    /// formula(value of base_snapshot at the end of the transaction)
+    SnapshotDerived {
+        base_snapshot: AggregatorID,
         formula: SnapshotToStringFormula,
     },
 }
 
-impl AggregatorChange {
-    /// Returns the result of applying the AggregatorState to the base value
-    /// Returns error if the aggregator behavior based on the
-    /// speculative_base_value isn't consistent with the provided base value.
-    pub fn apply_aggregator_change_to(&self, base: u128) -> PartialVMResult<u128> {
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum AggregatorChange {
```

Comment:
Trying to understand the design pattern here. Why is it better to divide AggregatorChange into AggregatorValue and AggregatorApplyChange, instead of having it as one piece as before?

Reply:
afterwards we handle Create and Apply very differently. And for each there is a lot of commonality for each.

For example, I can then add methods to AggregatorApplyChange - get_apply_base_id and apply_to_base  - and then treat all Apply's the same.  
I can utilize it directly in versioned_aggregators - and you can see how simpler things are there, i.e. we have:
Bypass(AggregatorApplyChange) (instead of 3 different bypasses)
similarly, we have only two:
    Value(AggregatorValue, Option<AggregatorApplyChange>),
    Apply(AggregatorApplyChange),
instead of 6 I would need to create otherwise.